### PR TITLE
Add number of simulations per server

### LINKED ISSUE #172: Add number of simulations per server to `Additive` init method
### 📝 Description of the feature

Add number of simulations per server to `Additive` init method to allow the client to perform more detailed load balancing when running on localhost or in pypim environment.

### 💡 Steps for implementing the feature

_No response_

### 🔗 Useful links and references

_No response_

### PR DESCRIPTION
**Pull request by pkrull-ansys**
closes #172

### CODE REVIEW

**Review by klmcadams — APPROVED** [review]
Looks good, nice examples!

File: src/ansys/additive/core/additive.py [hunk 1]
```diff
@@ -55,6 +55,15 @@
 class Additive:
     """Provides the client interface to one or more Additive services.
 
+    In a typical cloud environment, a single Additive service with load balancing and
+    auto-scaling is used. The ``Additive`` client connects to the service via a
+    single connection. However, for atypical environments or when running on localhost,
+    the ``Additive`` client can perform crude load balancing by connecting to multiple
+    servers and distributing simulations across them. The ``server_connections``,
+    ``nsims_per_server``, and ``nservers`` parameters can be used to control the
+    number of servers to connect to and the number of simulations to execute on each
```
Comment (julieatansys) [inline src/ansys/additive/core/additive.py:64]:
```suggestion
    number of servers to connect to and the number of simulations to run on each
```

File: src/ansys/additive/core/additive.py [hunk 2]
```diff
@@ -66,21 +75,48 @@ class Additive:
         ``server_channels`` or ``channel`` parameters is other than ``None``.
     port: int, default: 50052
         Port number to use when connecting to the server.
+    nsims_per_server: int, default: 1
+        Number of simultaneous simulations to execute on each server. Each simulation
```
Comment (julieatansys) [inline src/ansys/additive/core/additive.py:79]:
```suggestion
        Number of simultaneous simulations to run on each server. Each simulation
```

File: src/ansys/additive/core/additive.py [hunk 3]
```diff
@@ -158,14 +196,25 @@ def _connect_to_servers(
 
         return connections
 
+    @property
+    def nsims_per_server(self) -> int:
+        """Number of simultaneous simulations to execute on each server."""
```
Comment (julieatansys) [inline src/ansys/additive/core/additive.py:201]:
```suggestion
        """Number of simultaneous simulations to run on each server."""
```

File: src/ansys/additive/core/additive.py [hunk 4]
```diff
@@ -158,14 +196,25 @@ def _connect_to_servers(
...
+        return self._nsims_per_server
+
+    @nsims_per_server.setter
+    def nsims_per_server(self, value: int) -> None:
+        """Set the number of simultaneous simulations to execute on each server."""
```
Comment (julieatansys) [inline src/ansys/additive/core/additive.py:206]:
```suggestion
        """Set the number of simultaneous simulations to run on each server."""
```

File: src/ansys/additive/core/additive.py [hunk 5]
```diff
@@ -55,6 +55,15 @@
 class Additive:
     """Provides the client interface to one or more Additive services.
 
+    In a typical cloud environment, a single Additive service with load balancing and
+    auto-scaling is used. The ``Additive`` client connects to the service via a
+    single connection. However, for atypical environments or when running on localhost,
+    the ``Additive`` client can perform crude load balancing by connecting to multiple
+    servers and distributing simulations across them. The ``server_connections``,
+    ``nsims_per_server``, and ``nservers`` parameters can be used to control the
```
Comment (julieatansys) [inline src/ansys/additive/core/additive.py:63]:
```suggestion
    `, ``nservers``, and ``nsims_per_server` parameters can be used to control the
```
Perhaps reverse the order in which the parameters are listed to match their descriptions later in the sentence.

File: src/ansys/additive/core/additive.py [hunk 6]
```diff
@@ -66,21 +75,48 @@ class Additive:
         ``server_channels`` or ``channel`` parameters is other than ``None``.
     port: int, default: 50052
         Port number to use when connecting to the server.
+    nsims_per_server: int, default: 1
+        Number of simultaneous simulations to run on each server. Each simulation
+        requires a license checkout. If a license is not available, the simulation will
+        fail.
```
Comment (PipKat) [inline src/ansys/additive/core/additive.py:81]:
```suggestion
        requires a license checkout. If a license is not available, the simulation
        fails.
```

File: src/ansys/additive/core/additive.py [hunk 7]
```diff
@@ -66,21 +75,48 @@ class Additive:
...
     nservers: int, default: 1
-        Number of Additive servers to start and connect to.
-        ``nservers`` instances of Additive server will be started. For this to work,
-        the Additive portion of the Ansys Structures package must be installed.
-        This parameter is ignored if the ``server_connections``, ``channel``, or ``host``
-        parameter is other than ``None``.
+        Number of Additive servers to start and connect to. This parameter is only
+        applicable in `PyPIM`_ enabled cloud environments and on localhost. For
```
Comment (PipKat) [inline src/ansys/additive/core/additive.py:84]:
```suggestion
        applicable in `PyPIM`_-enabled cloud environments and on localhost. For
```

File: src/ansys/additive/core/additive.py [hunk 8]
```diff
@@ -55,6 +55,15 @@
 class Additive:
     """Provides the client interface to one or more Additive services.
 
+    In a typical cloud environment, a single Additive service with load balancing and
+    auto-scaling is used. The ``Additive`` client connects to the service via a
+    single connection. However, for atypical environments or when running on localhost,
+    the ``Additive`` client can perform crude load balancing by connecting to multiple
+    servers and distributing simulations across them. The ``server_connections``,
+    `, ``nservers``, and ``nsims_per_server` parameters can be used to control the
```
Comment (PipKat) [inline src/ansys/additive/core/additive.py:63]:
```suggestion
    servers and distributing simulations across them. The ``server_connections``,
    ``nservers``, and ``nsims_per_server` parameters can be used to control the
```

File: src/ansys/additive/core/additive.py [hunk 9]
```diff
@@ -66,21 +75,48 @@ class Additive:
...
-        Number of Additive servers to start and connect to.
-        ``nservers`` instances of Additive server will be started. For this to work,
-        the Additive portion of the Ansys Structures package must be installed.
-        This parameter is ignored if the ``server_connections``, ``channel``, or ``host``
-        parameter is other than ``None``.
+        Number of Additive servers to start and connect to. This parameter is only
+        applicable in `PyPIM`_ enabled cloud environments and on localhost. For
+        this to work on localhost, the Additive portion of the Ansys Structures
+        package must be installed. This parameter is ignored if the ``server_connections``,
+        or ``host`` parameter is other than ``None``.
```
Comment (PipKat) [inline src/ansys/additive/core/additive.py:87]:
```suggestion
        package must be installed. This parameter is ignored if the ``server_connections``
        parameter or ``host`` parameter is other than ``None``.
```

File: src/ansys/additive/core/additive.py [hunk 10]
```diff
@@ -66,21 +75,48 @@ class Additive:
...
     product_version: str
         Version of the Ansys product installation in the form ``"YYR"``, where ``YY``
         is the two-digit year and ``R`` is the release number. For example, the release
         2024 R1 would be specified as ``241``. This parameter is only applicable in
-        PyPIM environments and on localhost.
+        `PyPIM`_ enabled cloud environments and on localhost.
```
Comment (PipKat) [inline src/ansys/additive/core/additive.py:92]:
```suggestion
        `PyPIM`_-enabled cloud environments and on localhost.
```

File: src/ansys/additive/core/additive.py (hunk 8 above)
Comment (julieatansys) [inline src/ansys/additive/core/additive.py:63]:
```suggestion
     ``nservers``, and ``nsims_per_server` parameters can be used to control the
```
I think we had some extra tic marks

File: src/ansys/additive/core/additive.py [hunk 11]
```diff
@@ -66,21 +75,48 @@ class Additive:
...
     log_level: str, default: "INFO"
         Minimum severity level of messages to log.
     log_file: str, default: ""
         File name to write log messages to.
+
+    Examples
+    --------
+    Connecting to a list of servers. Multiple connections to the same host are permitted.
```
Comment (PipKat) [inline src/ansys/additive/core/additive.py:100]:
```suggestion
    Connect to a list of servers. Multiple connections to the same host are permitted.
```

File: src/ansys/additive/core/additive.py [hunk 12]
```diff
@@ -66,21 +75,48 @@ class Additive:
...
+
+    >>> additive = Additive(server_connections=["localhost:50052", "localhost:50052", "myserver:50052"])
+
+    Connecting to a single server using the host name and port number.
```
Comment (PipKat) [inline src/ansys/additive/core/additive.py:104]:
```suggestion
    Connect to a single server using the host name and port number.
```

File: src/ansys/additive/core/additive.py [hunk 13]
```diff
@@ -66,21 +75,48 @@ class Additive:
...
+
+    >>> additive = Additive(host="additive.ansys.com", port=12345)
+
+    Start and connect to two servers on localhost or in a
+    `PyPIM`_ enabled cloud environment. Allow each server to run two
```
Comment (PipKat) [inline src/ansys/additive/core/additive.py:109]:
```suggestion
    `PyPIM`_-enabled cloud environment. Allow each server to run two
```

File: src/ansys/additive/core/additive.py [hunk 14]
```diff
@@ -66,21 +75,48 @@ class Additive:
...
+    simultaneous simulations.
+
+    >>> additive = Additive(nsims_per_server=2, nservers=2)
+
+    Start a single server on localhost or in a `PyPIM`_ enabled cloud environment.
```
Comment (PipKat) [inline src/ansys/additive/core/additive.py:114]:
```suggestion
    Start a single server on localhost or in a `PyPIM`_-enabled cloud environment.
```

File: src/ansys/additive/core/additive.py [hunk 15]
```diff
@@ -158,14 +196,25 @@ def _connect_to_servers(
...
+    @property
+    def nsims_per_server(self) -> int:
+        """Number of simultaneous simulations to run on each server."""
+        return self._nsims_per_server
+
+    @nsims_per_server.setter
+    def nsims_per_server(self, value: int) -> None:
+        """Set the number of simultaneous simulations to run on each server."""
+        if value < 1:
+            raise ValueError("Number of simulations per server must be greater than zero")
```
Comment (PipKat) [inline src/ansys/additive/core/additive.py:208]:
```suggestion
            raise ValueError("Number of simulations per server must be greater than zero.")
```
Message strings should have concluding punctuation.

File: src/ansys/additive/core/additive.py [hunk 16]
```diff
@@ -158,14 +196,25 @@ def _connect_to_servers(
...
+        self._nsims_per_server = value
+
     def about(self) -> None:
         """Print information about the client and server."""
         print(f"Client {__version__}, API version: {api_version}")
         if self._servers is None:
-            print("Not connected to server")
+            print("Not connected to a server")
```
Comment (PipKat) [inline src/ansys/additive/core/additive.py:215]: Consider changing to "Client is not connected to a server."
Using short, concise sentences is preferred.

File: src/ansys/additive/core/server_connection/server_connection.py [hunk 17]
```diff
@@ -65,9 +65,10 @@ class ServerConnection:
     """Provides connection to Additive server.
 
     If neither ``channel`` nor ``addr`` are provided, an attempt will be
```
Comment (PipKat) [inline src/ansys/additive/core/server_connection/server_connection.py:67]:
```suggestion
    If neither ``channel`` nor ``addr`` are provided, an attempt is
```

File: src/ansys/additive/core/server_connection/server_connection.py [hunk 18]
```diff
@@ -80,7 +81,8 @@ class ServerConnection:
         Version of the Ansys product installation in the form ``"YYR"``, where ``YY``
         is the two-digit year and ``R`` is the release number. For example, the release
         2024 R1 would be specified as ``241``. This parameter is only applicable in
-        PyPIM environments and on localhost.
+        `PyPIM <https://pypim.docs.pyansys.com/version/stable/index.html>`_
```
Comment (PipKat) [inline src/ansys/additive/core/server_connection/server_connection.py:84]:
```suggestion
        `PyPIM <https://pypim.docs.pyansys.com/version/stable/index.html>`_-
```

**Review by PipKat — APPROVED** [review]
Suggested some very minor edits.

File: src/ansys/additive/core/additive.py (hunk 8 above)
Comment (pkrull-ansys) [reply src/ansys/additive/core/additive.py:63]: duplicate of Kathy's suggestion.

File: src/ansys/additive/core/additive.py (hunk 10 above)
Comment (PipKat) [inline src/ansys/additive/core/additive.py]:
```suggestion
        `PyPIM`_-enabled cloud environments and on localhost.
```

File: src/ansys/additive/core/additive.py [hunk 19]
```diff
@@ -66,21 +75,48 @@ class Additive:
...
     product_version: str
         Version of the Ansys product installation in the form ``"YYR"``, where ``YY``
         is the two-digit year and ``R`` is the release number. For example, the release
         2024 R1 would be specified as ``241``. This parameter is only applicable in
-        PyPIM environments and on localhost.
+        `PyPIM`_ enabled cloud environments and on localhost.
     log_level: str, default: "INFO"
         Minimum severity level of messages to log.
     log_file: str, default: ""
         File name to write log messages to.
```
[the remark this answers is not in the document]
Comment (pkrull-ansys) [reply src/ansys/additive/core/additive.py:96]: Good question! It's an error but I don't have any checking for it at present.

File: src/ansys/additive/core/server_connection/server_connection.py (hunk 18 above)
Comment (pkrull-ansys) [reply src/ansys/additive/core/server_connection/server_connection.py]: done

File: src/ansys/additive/core/additive.py (hunk 10 above)
Comment (pkrull-ansys) [reply src/ansys/additive/core/additive.py]: done

**Pull request merged**